### ISSUE TITLE
SSE support

### ISSUE THREAD

**Issue by dberecz** [dapr/dapr#3830]
## In what area(s)?

 /area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release

## Ask your question here

Hi all, I'm struggling with using dapr in case of Server Sent Events. I have an HTTP endpoint which is SSE configured, and can also return a plain JSON, depending on `accept` header. If I call the endpoint directly in case of SSE, I get the correct streamed responses. If I call it through dapr, I get no response back, see below. Does it mean SSE is not supported or do I need to do some additional configuration? Thanks a lot!

`curl -H 'accept: text/event-stream' -N 'http://localhost:3500/v1.0/invoke/denoapp/method/schedule'`--> No response
`curl -H 'accept: application/json' -N 'http://localhost:3500/v1.0/invoke/denoapp/method/schedule'`--> Correct response
`curl -H 'accept: text/event-stream' -N 'http://denoapp:80/schedule'` --> Correct response (calling directly, without dapr)

**Comment by daixiang0** [comment]
Looks it is a new feature, could you provide more details about SSE and use cases?

**Comment by xinyii** [comment]
Node.js SSE Server example:

```js
const express = require('express')

const app = express()

app.get('/countdown', function(req, res) {
  res.writeHead(200, {
    'Content-Type': 'text/event-stream',
    'Cache-Control': 'no-cache',
    'Connection': 'keep-alive'
  })
  countdown(res, 10)
})

function countdown(res, count) {
  res.write("data: " + count + "\n\n")
  if (count)
    setTimeout(() => countdown(res, count-1), 1000)
  else
    res.end()
}

app.listen(3000, () => console.log('SSE app listening on port 3000!'))
```

As shown above, the response will not be returned at once, but written in multiple times, and even the response will never end.



See also:

- <https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events>
- <https://html.spec.whatwg.org/multipage/server-sent-events.html>

**Comment by dberecz** [comment]
We would like to have a server (with dapr) providing an SSE endpoint (similarly as @xinyii described above) so that a frontend can subscribe and receive events to update the UI. Similar to websockets but only one-way communication. Does it sound reasonable?

**Comment by danielberecz** [comment]
@daixiang0 Would it make sense to pin this issue to avoid being closed?

**Comment by dberecz** [comment]
@daixiang0 Would it make sense to pin this issue to avoid being closed?

**Comment by jjcollinge** [comment]
@dberecz just trying to clarify your use case - your frontend is calling your backend service (with the SSE endpoint) via Dapr rather than directly? Where does your frontend run (browser, native, streamed, etc.)? Are you running a Dapr instance client side with your frontend or have you exposed the backend service's Dapr API?

**Comment by EashShow** [comment]
I faced the same problem #4336

**Comment by sstunkel** [comment]
> @ItalyPaleAle is the work you're doing on streaming going to enable this?

Just as a side note, we're doing a lot of streming too and if there's perhaps an upcoming landmine we'd love to know what it is =)

**Comment by ItalyPaleAle** [comment]
@sstunkel I'm not sure I fully understand however how you're using Dapr with SSE with GraphQL. Do you have a service that exposes the GraphQL endpoint (let's call it G), then a "front-end" service A which talks to G through Dapr, and then clients connect to A?

I don't know if this is a supported scenario with Dapr today and it may be something we will support only very much down the road, for HTTP.

The work I'm doing on streaming is a bit different at the moment. #5170 (which will not be merged as-is) is mostly focused on chunking (and sending chunks over a stream) messages that are related to HTTP service invocation, so sidecar-to-sidecar communication is stream-based. The main problem I've been trying to solve was sending large payloads between Dapr sidecars efficiently, so Dapr never has to buffer the entire message in memory first (which adds latency and also memory pressure).

However, with #5170 I'm still primarily concerned with "simple RPCs", meaning where app A invokes app B, and app B responds "right away". Support for long-lived streams is not in scope of that PR.

For your use case, I would recommend using gRPC proxying with Dapr, which is optimized for long-lived streams.

**Comment by ItalyPaleAle** [comment]
> Or are you referring specifically that SSE in this scenario isn't supported?

I am saying that SSE is not a supported scenario in this case.

> Just want to make sure I understand correctly - if a service tries to stream from another service using chunked-transfer encoding of HTTP/1.1, its currently getting buffered entirely by the sidecar? But your PR seems to partially fix that?

Correct. Currently, when using HTTP, requests and responses are fully buffered in memory by Dapr. My PR will address that, but it's still not going to enable long-lived connections (at least, that's not an officially-supported scenario in #5170)

Curious to why you are using Dapr with your ingress controller by the way, rather than just making nginx invoke the app?

**Comment by olitomlinson** [comment]
> Curious to why you are using Dapr with your ingress controller by the way, rather than just making nginx invoke the app?

It's a strategy recommended by the dapr maintainers. Particularly for e2e mTLS. 

Does having a dapr-ised ingress create additional complexities for supporting SSE?

**Comment by ItalyPaleAle** [comment]
> Does having a dapr-ised ingress create additional complexities for supporting SSE?

Yes at this stage. I would not recommend using SSE through Dapr via HTTP service invocation. If you can wrap it in gRPC proxied invocations, then it'd be fine. But we can't currently support the scenario you're looking into.

**Comment by olitomlinson** [comment]
> if you can wrap it in gRPC proxied invocations, then it'd be fine.

Sorry, just to be clear, are you saying that if we wrapped the call between nginx sidecar and the target service in gRPC everything would just work? with no further changes required to support SSE on our side or on dapr runtime?

**Or** 

Are you saying wrapping the call between nginx and the target service in gRPC, would help to make SSE possible, but there would still need to be further work in dapr runtime to support SSE?

Thanks @ItalyPaleAle

**Comment by olitomlinson** [comment]
@dberecz 

Good news - SSE is now supported with Service Invoke as of 1.17.

Please can you retest your reproduction using the [latest 1.17 RC ](https://github.com/dapr/dapr/releases)runtime release. If you have problems please reach out. Thank you

Cc @javier-aliaga
